shared/oae/api/oae.api.authentication.js

```diff
@@ -13,74 +13,231 @@
         }
 
         // Twitter authentication
-        if (configAPI.getValue('oae-authentication', 'twitter', 'enabled')) {
-            enabledStrategies['twitter'] = {'url': '/api/auth/twitter'};
+        if (configAPI.getValue('oae-authentication', STRATEGY_TWITTER, 'enabled')) {
+            enabledStrategies[STRATEGY_TWITTER] = {
+                'icon': 'twitter',
+                'name': i18nAPI.translate('__MSG__' + contextLabel + '_WITH_TWITTER__'),
+                'url': '/api/auth/twitter'
+            };
         }
 
         // Local authentication
-        if (configAPI.getValue('oae-authentication', 'local', 'enabled')) {
-            enabledStrategies['local'] = {'url': '/api/auth/login'};
+        if (configAPI.getValue('oae-authentication', STRATEGY_LOCAL, 'enabled')) {
+            enabledStrategies[STRATEGY_LOCAL] = {'url': '/api/auth/login'};
         }
 
         return enabledStrategies;
     };
 
     /**
+     * Determine if there is a login redirect url available for the current page
+     *
+     * @param  {String}     [location]  The url to use to check for the redirect url. If unspecified, the current window location will be used
+     * @return {String}                 The login redirect url, if any
+     */
+    var getLoginRedirectUrl = exports.getLoginRedirectUrl = function(location) {
+        return utilAPI.url(location).param('url');
+    };
+
+    /**
+     * Perform a login with an external authentication strategy
+     *
+     * @param  {String}     strategyId              The id of the strategy with which to authenticate. It must be a strategy that is enabled in the list of `enabledExternalStrategies` from the strategy info
+     * @param  {Object}     [opts]                  Optional arguments
+     * @param  {String}     [opts.redirectUrl]      The redirect url to follow after authentication success
+     * @param  {String}     [opts.invitationToken]  The invitation token from which the login originates, if any
+     */
+    var externalLogin = exports.externalLogin = function(strategyId, opts) {
+        if (!strategyId) {
+            throw new Error('A valid strategy id should be provided');
+        }
+
+        // Ensure we were provided an enabled external strategy
+        var strategyInfo = getStrategyInfo();
+        var strategy = strategyInfo.enabledExternalStrategies[strategyId];
+        if (!strategy) {
+            throw new Error('Strategy id must be an enabled external strategy');
+        }
+
+        opts = opts || {};
+
+        // Use the `auth.html` `authExternalButton` macro to create a form that performs this
+        // authentication
+        var $template = $('<div style="display: none";>${authExternalButton(strategyId, strategy, opts)}</div>');
+        var form = utilAPI.template().render($template, {
+            'strategyId': strategyId,
+            'strategy': strategyInfo.enabledExternalStrategies[strategyId],
+            'opts': {
+                'data': {
```

Comment:
In this case, `data` refers to extra data parameters to add to the login form for the authentication request -- it will take any parameter added here. I'd like to keep it as a recognized `data` parameter, so that more `opts` options can be added without unintentionally (and probably unknowingly) adding more data to the login request which could eventually result in a bug.

For the same reason, I'd like to individually pick the options off the `opts` parameters as it is being done now since adding more opts parameters could unintentionally result in more data parameters in the form submission and subsequently cause bugs.